tests/functional/testWebExtensionApi.js

```diff
@@ -0,0 +1,89 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const assert = require('assert');
+const vpn = require('./helper.js');
+const queries = require('./queries.js');
+
+const {
+  sentToClient,
+  connectExtension,
+  getMessageStream,
+  ExtensionMessage,
+  readResponseOfType
+} = require('./utils/webextension.js')
+
+
+describe('WebExtension API', function() {
+  this.timeout(60000);
+  this.ctx.authenticationNeeded = true;
+
+  it('A Webextension can query the Status of the VPN', async () => {
+    const sock = await connectExtension();
+    const messagePipe = getMessageStream(sock);
+    sentToClient(new ExtensionMessage('status'), sock);
+    await readResponseOfType('status', messagePipe);
+    sock.destroy();
+  });
+  it('A Webextension can activate the VPN', async () => {
+    const sock = await connectExtension();
+    const messagePipe = getMessageStream(sock);
+    sentToClient(new ExtensionMessage('activate'), sock);
+
+    await vpn.waitForCondition(async () => {
+      sentToClient(new ExtensionMessage('status'), sock);
+      const msg = await readResponseOfType('status', messagePipe);
+      return msg.status.vpn === 'StateOn';
+    });
+    const currentstate =
+        await vpn.getMozillaProperty('Mozilla.VPN', 'VPNController', 'state');
+    assert(currentstate === 'StateOnPartial')
+    sock.destroy();
+  });
+  it('A Webextension can deactivate the VPN if it Self Activated', async () => {
+    const sock = await connectExtension();
+    const messagePipe = getMessageStream(sock);
+    sentToClient(new ExtensionMessage('activate'), sock);
+    await readResponseOfType('activate', messagePipe);
+    await vpn.waitForCondition(async () => {
+      sentToClient(new ExtensionMessage('status'), sock);
+      const msg = await readResponseOfType('status', messagePipe);
+      return msg.status.vpn === 'StateOn';
+    });
+    sentToClient(new ExtensionMessage('deactivate'), sock);
+    await vpn.waitForCondition(async () => {
+      sentToClient(new ExtensionMessage('status'), sock);
+      const msg = await readResponseOfType('status', messagePipe);
+      return msg.status.vpn === 'StateOff';
+    });
+    sock.destroy();
+  });
+
+  it('A Webextension can NOT deactivate the VPN if it was activated in Client',

```

Comment:
Related to other comment: Eventually we _will_ allow deactivating the vpn for Firefox in this situation right? This may be a terrible idea but I'm almost wondering if we'll want another state machine for the extension to make keeping track of situations like _client is on but Firefox VPN is off_ a bit easier. Otherwise if we do allow the user to remove Firefox traffic from the tunnel when the vpn was activated from the client, but tell the extension that the VPN is 'on', how will the extension know that Firefox is excluded/"off"?

Reply:
The Webextension state machine is going to be diffrent from the client. 
If full device protection is enabled, the extension can use the localproxy to "exlude firefox"
-> if it's in partialOn we can just not use a proxy. 

Thinking about this more, i think we might need to tell the extension actually if it is partially or full-device on. 
But in no situation we need the extension to downgrade from full device protection :)